File: app/src/main/java/com/rohitsuratekar/NCBSinfo/activities/AddTripsFragment.java
package com.rohitsuratekar.NCBSinfo.activities;


import android.app.TimePickerDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TimePicker;

import com.rohitsuratekar.NCBSinfo.R;
import com.secretbiology.helpers.general.Log;

import java.util.Calendar;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by Rohit on 7/2/2017.
 */

public class AddTripsFragment extends Fragment {

    @BindView(R.id.et_trip_recycler)
    RecyclerView recyclerView;
    private AddTripsAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.edit_transport_trips, container, false);
        ButterKnife.bind(this, rootView);
        adapter = new AddTripsAdapter();
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);
        recyclerView.setHasFixedSize(true);
        recyclerView.addItemDecoration(new SwipeItemDecorator());
        ItemTouchHelper mItemTouchHelper = new ItemTouchHelper(TouchHelper.get(getContext(), recyclerView));
        mItemTouchHelper.attachToRecyclerView(recyclerView);
        adapter.setUndoOn(true);
        return rootView;
    }

    @OnClick(R.id.et_pick_time)
    public void openPicker() {
        Calendar calendar = Calendar.getInstance();
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);
        TimePickerDialog mTimePicker;
        mTimePicker = new TimePickerDialog(getContext(), new TimePickerDialog.OnTimeSetListener() {
            @Override
            public void onTimeSet(TimePicker timePicker, int selectedHour, int selectedMinute) {
                Log.inform();
            }
        }, hour, minute, false);//Yes 24 hour time
        mTimePicker.show();
    }
}

File: app/src/main/java/com/rohitsuratekar/NCBSinfo/fragments/AddTripsFragment.kt
package com.rohitsuratekar.NCBSinfo.fragments


import android.app.TimePickerDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.rohitsuratekar.NCBSinfo.R
import com.rohitsuratekar.NCBSinfo.adapters.EditTransportTripAdapter
import com.rohitsuratekar.NCBSinfo.common.*
import com.rohitsuratekar.NCBSinfo.models.EditFragment
import kotlinx.android.synthetic.main.fragment_add_trips.*
import java.util.*

class AddTripsFragment : EditFragment() {

    private lateinit var adapter: EditTransportTripAdapter
    private var itemList = mutableListOf<String>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_add_trips, container, false)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter = EditTransportTripAdapter(itemList)
        adapter.setUndoOn(false)
        adapter.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onChanged() {
                super.onChanged()
                sharedModel.updateTrips(itemList)
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                super.onItemRangeRemoved(positionStart, itemCount)
                itemList.sort()
                adapter.notifyDataSetChanged()
                sharedModel.updateTrips(itemList)
                sharedModel.updateTripSelection(false)
                checkEmptyList()
            }
        })
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        callback?.hideProgress()
        sharedModel.updateReadState(Constants.EDIT_TRIPS)
        callback?.setFragmentTitle(R.string.et_add_trips)
        et_trip_add_fab.setOnClickListener { selectTime() }
        et_trip_next.setOnClickListener { callback?.navigate(Constants.EDIT_START_TRIP) }
        et_trip_previous.setOnClickListener { callback?.navigateWithPopback() }
        setUpRecycler()
    }

    private fun setUpRecycler() {
        checkOldData()
        et_trip_recycler.layoutManager = LinearLayoutManager(context)
        et_trip_recycler.adapter = adapter
        et_trip_recycler.setHasFixedSize(true)
        et_trip_recycler.addItemDecoration(SwipeItemDecorator())
        val mItemTouchHelper = ItemTouchHelper(TouchHelper.get(context!!, et_trip_recycler))
        mItemTouchHelper.attachToRecyclerView(et_trip_recycler)
        checkEmptyList()
    }

    private fun checkEmptyList() {
        if (itemList.isEmpty()) {
            et_trip_holder.hideMe()
        } else {
            et_trip_holder.showMe()
        }
    }

    private fun getFormattedDate(hour: Int, minute: Int): String {
        var returnString = hour.toString()
        if (returnString.length == 1) {
            returnString = "0$returnString"
        }
        returnString = if (minute.toString().length == 1) {
            "$returnString:0$minute"
        } else {
            "$returnString:$minute"
        }
        return returnString
    }

    private fun selectTime() {
        val cal = Calendar.getInstance()
        TimePickerDialog(context!!, TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
            val t = getFormattedDate(hourOfDay, minute)
            if (t in itemList) {
                context?.toast(getString(R.string.et_same_trip_warning, convertTimeFormat(t)))
            } else {
                itemList.add(t)
                itemList.sort()
                adapter.notifyDataSetChanged()
                checkEmptyList()
                sharedModel.updateTripSelection(false)
            }
        }, cal.get(Calendar.HOUR_OF_DAY), cal.get(Calendar.MINUTE), false).show()
    }

    private fun checkOldData() {

        sharedModel.tripList.value?.let {
            itemList.clear()
            itemList.addAll(it)
            adapter.notifyDataSetChanged()
            checkEmptyList()
        }
    }

}
